src/Suave.Tests/Owin.fs

```diff
@@ -148,5 +159,177 @@ let endToEnd =
+
+      runWithConfig composedApp |> reqResp Types.GET "/" "" None None DecompressionMethods.GZip id asserts
+
+    testCase "Completed Async signals completion with status code and headers" <| fun _ ->
+      let noContent (env : OwinEnvironment) =
+        env.[OwinConstants.responseStatusCode] <- box 204
+        let responseHeaders : IDictionary<string, string[]> = unbox env.[OwinConstants.responseHeaders]
+        responseHeaders.["Content-Type"] <- [| "text/plain; charset=utf-8" |]
+        async.Return ()
+
+      let composedApp =
+        OwinApp.ofApp noContent
+
+      let asserts (result : HttpResponseMessage) =
+        eq "Http Status Code" HttpStatusCode.NoContent result.StatusCode
+        eq "Reason Phrase set by server" "No Content" result.ReasonPhrase
+        eq "Content-Type" "text/plain; charset=utf-8" (result.Content.Headers.ContentType.ToString())
+
+      runWithConfig composedApp |> reqResp Types.GET "/" "" None None DecompressionMethods.GZip id asserts
+
+    testCase "Custom reason phrase" <| fun _ ->
+      let noContent (env : OwinEnvironment) =
+        env.[OwinConstants.responseStatusCode] <- box 204
+        env.[OwinConstants.responseReasonPhrase] <- box "Nothing to see here"
+        async.Return ()
+
+      let composedApp =
+        OwinApp.ofApp noContent
+
+      let asserts (result : HttpResponseMessage) =
+        eq "Http Status Code" HttpStatusCode.NoContent result.StatusCode
+        eq "Reason Phrase" "Nothing to see here" result.ReasonPhrase
+
+      runWithConfig composedApp |> reqResp Types.GET "/" "" None None DecompressionMethods.GZip id asserts
+
+    testCase "OWIN middleware runs as a Suave app" <| fun _ ->
+      // This test case exists to show that a middleware will mount into Suave as an application.
+      let noContent = OwinMidFunc(fun next -> OwinAppFunc(fun env ->
+        env.[OwinConstants.responseStatusCode] <- box 204
+        Threading.Tasks.Task.FromResult() :> Threading.Tasks.Task
+        ))
+
+      let composedApp =
+        OwinApp.ofMidFunc noContent
+
+      let asserts (result : HttpResponseMessage) =
+        eq "Http Status Code" HttpStatusCode.NoContent result.StatusCode
+        eq "Reason Phrase set by server" "No Content" result.ReasonPhrase
+
+      runWithConfig composedApp |> reqResp Types.GET "/" "" None None DecompressionMethods.GZip id asserts
+
+    testCase "Composed OWIN security middleware and app" <| fun _ ->
+      let noContent = OwinAppFunc(fun env ->
+        env.[OwinConstants.responseStatusCode] <- box 204
+        Threading.Tasks.Task.FromResult() :> Threading.Tasks.Task
+        )
+      
+      let basicAuthMidFunc = OwinMidFunc(fun next -> OwinAppFunc(fun env ->
+        let requestHeaders : IDictionary<string, string[]> = unbox env.[OwinConstants.requestHeaders]
+        // NOTE: currently fails b/c headers are not using StringComparer.OrdinalIgnoreCase.
+        let authHeader = requestHeaders.["Authorization"].[0].Split([|' '|])
```

Comment:
Since I didn't state it explicitly, my understanding as to why this test fails is that I used `"Authorization"` rather than `"authorization"`, as the `KeyNotFoundException` is thrown, and `"authorization"` exists in the headers dictionary.